drivers/QSPI.h

```diff
@@ -222,13 +234,15 @@ class QSPI : private NonCopyable<QSPI> {
     int _mode; //SPI mode
     bool _initialized;
     PinName _qspi_io0, _qspi_io1, _qspi_io2, _qspi_io3, _qspi_clk, _qspi_cs; //IO lines, clock and chip select
+    const qspi_pinmap_t *_explicit_pinmap;
```

Comment:
Are you requiring the passed-in pinmap object to have lifetime extending beyond the constructor call? That's not specified.

~~Looking at the code, I don't think you are, so this doesn't have to be a member. You could just pass the pinmap to the `_initialize_direct`. Will save 4 bytes of RAM.~~ (Nope - you should be calling `_initialize_direct other places)

I realise this model comes from the `SPI` code, which _does_ require the pinmap to persist for later re-init. That one does need that specified. Maybe the same should be specified for all classes to allow suspend/resume in future, even if they don't do it now.

Reply:
Pondering the persistence thing - this means that

    SPI(get_spi_pinmap(A,B,C,D))

which I've seen suggested is actually illegal. That would be retaining a pointer to a temporary.

It's unfortunate that the compiler accepts it. You're allowed to pass a temporary `X` to a `const X &`

Two suggestions to prevent that here: https://stackoverflow.com/questions/40790126/prevent-passing-temporary-for-const-ref-parameter

1. Delete the overload with r-value parameter: `SPI(const spi_pinmap_t &&) = delete`
1. Use const pointer: ` SPI(const spi_pinmap_t *)`

Not sure it's too worth worrying about.

Reply:
The mentioned form `SPI(get_spi_pinmap(A,B,C,D))` is actually not suggested. According to the design doc:
> It is strongly suggested that the result of `get_xxx_pinmap()` calls should normally be used in explicitly constant expression contexts like initializing a constexpr variable, to guarantee that the tables are not pulled in.

I think we should prevent passing of temporary objects. The first option looks more suitable for our case.

Reply:
I added 
`SPI(const spi_pinmap_t &&, PinName) = delete;`

And now when `SPI spi(get_spi_pinmap(PA_7, PA_6, PA_5, PA_4));` causes compilation error:

```
[Error] main.cpp@41,9: call to deleted constructor of 'mbed::SPI'
[ERROR] .\main.cpp:41:9: error: call to deleted constructor of 'mbed::SPI'
    SPI spi(get_spi_pinmap(PA_7, PA_6, PA_5, PA_4));
        ^   ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
./drivers/SPI.h:146:5: note: 'SPI' has been explicitly marked deleted here
    SPI(const spi_pinmap_t &&) = delete;
    ^
1 error generated.
```

This is what we wanted to achieve. The last question is if this solution should be added only to `SPI` and `QSPI` classes or to all peripheral constructors (which support explicit pinmap)?